### PR TITLE
docs: decision record about configuration injection

### PR DESCRIPTION
**Pull request by paullatzelsperger**
## What this PR changes/adds

Adds a decision record that outlines the configuration injection feature

## Linked Issue(s)

Closes # <-- _insert Issue number if one exists_

_Please be sure to take a look at the [contributing guidelines](https://github.com/eclipse-edc/.github/blob/main/CONTRIBUTING.md#submit-a-pull-request) and our [etiquette for pull requests](https://github.com/eclipse-edc/.github/blob/main/contributing/pr_etiquette.md)._

### CODE REVIEW

**Review by wolf4ood — APPROVED** [review]

**Review by jimmarino — COMMENTED** [review]
When we have errors reported, can we not use exceptions and stack traces? I would prefer a severe message containing the class requiring the injection point, field name, and expected type. That will be easier than reading stacktraces.

**Comment by paullatzelsperger** [comment]
> When we have errors reported, can we not use exceptions and stack traces? I would prefer a severe message containing the class requiring the injection point, field name, and expected type. That will be easier than reading stacktraces.

The runtime won't boot in the case a dependency or a mandatory config setting is not satisfied, so an exception is perfectly fine. In addition, this would impact areas of code outside of the scope of this DR, so that would be a separate discussion anyway.

**Comment by jimmarino** [comment]
> > When we have errors reported, can we not use exceptions and stack traces? I would prefer a severe message containing the class requiring the injection point, field name, and expected type. That will be easier than reading stacktraces.
> 
> The runtime won't boot in the case a dependency or a mandatory config setting is not satisfied, so an exception is perfectly fine, and the information you mentioned is indeed logged. In addition, changing the way dependency errors are reported would impact areas of code outside of the scope of this DR, so that would be a separate discussion anyway. My goal here was to align the error reporting of config injection with that of dependency injection.

That's not what I'm saying. When the missing config error is encountered, it should be reported to a context where it is collated, the loading for that extension is aborted, and loading then continues. After the phase is complete, the context should be checked, and the recorded errors (not exception) reported to the monitor. After the errors are reported, an exception can be thrown (or the runtime terminated).

File: docs/developer/decision-records/2024-11-06-configuration-injection/README.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+### Error reporting
+
+By piggy-backing on the dependency injection mechanism, errors are automatically reported in a collated way. So multiple
+unresolved config values would be reported as one:
+
+```
+Caused by: org.eclipse.edc.boot.system.injection.EdcInjectionException: The following injected fields were not provided or could not be resolved:
+Configuration value "fooBarBaz" of type class java.lang.Long (config key 'foo.bar.baz')
+Configuration object 'config' of type 'class org.eclipse.edc.some.FooBarExtension$FooBarConfig' in class class org.eclipse.edc.some.FooBarExtension
```
Comment (ndr-brt) [inline docs/developer/decision-records/2024-11-06-configuration-injection/README.md:133]:
```suggestion
Configuration object 'config' of type 'class org.eclipse.edc.some.FooBarExtension$FooBarConfig' in class org.eclipse.edc.some.FooBarExtension
```

**Review by ndr-brt — APPROVED** [review]

**Comment by ndr-brt** [comment]
> That's not what I'm saying. When the missing config error is encountered, it should be reported to a context where it is collated, the loading for that extension is aborted, and loading then continues. After the phase is complete, the context should be checked, and the recorded errors (not exception) reported to the monitor. After the errors are reported, an exception can be thrown (or the runtime terminated).

how could loading continues if an extension is aborted? other extensions could be dependent of that one so likely other kind of errors could appear.
As it is done for `@Injection` points, it should fail at the first inconsistency

File: docs/developer/decision-records/2024-11-06-configuration-injection/README.md (hunk 1 above)
Comment (paullatzelsperger) [reply docs/developer/decision-records/2024-11-06-configuration-injection/README.md:133]: copy paste...

**Comment by jimmarino** [comment]
> > That's not what I'm saying. When the missing config error is encountered, it should be reported to a context where it is collated, the loading for that extension is aborted, and loading then continues. After the phase is complete, the context should be checked, and the recorded errors (not exception) reported to the monitor. After the errors are reported, an exception can be thrown (or the runtime terminated).
> 
> how could loading continues if an extension is aborted? other extensions could be dependent of that one so likely other kind of errors could appear. As it is done for `@Injection` points, it should fail at the first inconsistency

The point of this is to collate all missing configuration points. This may need to perform config injection before service injection to achieve that. During config injection, if there are errors, all operations on the extension in question should cease and the booter moves to the next extension.

**Review by jimmarino — APPROVED** [review]

**Pull request merged**